File: 2023/02/python/cube_conundrum.py
def part_1(file: str):
    bag = {"red": 12, "green": 13, "blue": 14}

    with open(file, "r") as f:
        puzzle_contents = f.read().splitlines()

    possible_games = 0

    for line in puzzle_contents:
        game_num = int(line[: line.find(":")].split()[1])
        game_sets = line[line.find(":") + 1 :].split(";")
        game_set_num = 0
        is_possible = True
        while (game_set_num < len(game_sets)) & is_possible:
            cubes = game_sets[game_set_num].split(",")
            for cube in cubes:
                cube_color = cube.strip().split(" ")[1]
                cube_qty = int(cube.strip().split(" ")[0])
                if bag.get(cube_color, 0) < cube_qty:
                    is_possible = False
            game_set_num += 1
        possible_games += game_num if is_possible else 0
    return possible_games


def part_2(file: str):
    with open(file, "r") as f:
        puzzle_contents = f.read().splitlines()

    sum_of_power = 0
    for line in puzzle_contents:
        game_sets = line[line.find(":") + 1 :].split(";")
        game_set_num = 0
        max_so_far = {}
        while game_set_num < len(game_sets):
            cubes = game_sets[game_set_num].split(",")
            for cube in cubes:
                cube_color = cube.strip().split(" ")[1]
                cube_qty = int(cube.strip().split(" ")[0])
                if max_so_far.get(cube_color, 0) < cube_qty:
                    max_so_far[cube_color] = cube_qty
            game_set_num += 1
        power = 1
        for value in max_so_far.values():
            power *= value
        sum_of_power += power
    return sum_of_power


if __name__ == "__main__":
    puzzle_input = "day_02_puzzle_input.txt"
    with open(puzzle_input, "r") as f:
        puzzle_contents = f.read().splitlines()

    print("Answer to Part 1:", part_1(puzzle_contents))
    print("Answer to Part 2:", part_2(puzzle_contents))

File: 2023/02/typescript/cube-conundrum.ts
import { readFile } from "node:fs/promises";
import path from "node:path";

const CUBE_COLORS = ["red", "green", "blue"] as const;

type CubeColors = (typeof CUBE_COLORS)[number];
type CubeSet = [number, CubeColors];
type CubeGame = {
  gameId: number;
  draws: CubeSet[][];
};

function assertCubeColor(str: string | undefined): asserts str is CubeColors {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  if (!CUBE_COLORS.includes(str as any)) {
    throw new Error(`Not of type color: ${CUBE_COLORS.join(",")}`);
  }
}

export function parseInput(data: string): CubeGame[] {
  const games: CubeGame[] = [];
  for (const line of data.split("\n")) {
    // @ts-expect-error temp comment to ignore undefined error
    const gameId = Number(line.split(":")[0].split(" ")[1]);
    const rawGameValues = line.split(":")[1];

    if (rawGameValues !== undefined) {
      const draws: CubeSet[][] = [];
      for (const rawCubeSet of rawGameValues.split(";")) {
        const cubeSets: CubeSet[] = [];
        for (const rawCube of rawCubeSet.split(",")) {
          if (typeof rawCube === "string") {
            const rawCubeParts = rawCube.trim().split(" ");
            const numberOfCubes = Number(rawCubeParts[0]);
            const rawCubeColor = rawCubeParts[1];
            assertCubeColor(rawCubeColor);
            const cubeSet: CubeSet = [numberOfCubes, rawCubeColor];
            cubeSets.push(cubeSet);
          }
        }
        draws.push(cubeSets);
      }
      const cubeGame: CubeGame = {
        gameId: gameId,
        draws: draws,
      };
      games.push(cubeGame);
    }
  }
  return games;
}

export function part1(cubeGames: CubeGame[]): number {
  const NUMBER_OF_CUBES: Record<CubeColors, number> = {
    red: 12,
    green: 13,
    blue: 14,
  };

  function isLessThanTotalCubes(element: CubeSet[]) {
    return element.every((cubeSet) => {
      const [numberOfCubes, cubeColor] = cubeSet;
      return numberOfCubes <= NUMBER_OF_CUBES[cubeColor];
    });
  }

  const possibleGames = cubeGames.filter((game) =>
    game.draws.every(isLessThanTotalCubes),
  );

  return possibleGames.reduce((sum, game) => sum + game.gameId, 0);
}

export function part2(cubeGames: CubeGame[]): number {
  let result = 0;
  for (const game of cubeGames) {
    const maxNumberofCubes: Record<CubeColors, number> = {
      red: 0,
      green: 0,
      blue: 0,
    };
    for (const cubeSet of game.draws.flat()) {
      const [numberOfCubes, cubeColor] = cubeSet;
      if (numberOfCubes > maxNumberofCubes[cubeColor]) {
        maxNumberofCubes[cubeColor] = numberOfCubes;
      }
    }

    const power = Object.values(maxNumberofCubes).reduce((power, current) => {
      if (current > 0) {
        power *= current;
      }
      return power;
    });
    result += power;
  }
  return result;
}

if (import.meta.filename === process.argv[1]) {
  const inputFilePath = path.join(
    import.meta.dirname,
    "..",
    "data",
    "input.txt",
  );
  const data = await readFile(inputFilePath, "utf-8");
  console.log("Answer to part 1:", part1(parseInput(data)));
  console.log("Answer to part 2:", part2(parseInput(data)));
}
